### PR TITLE
Remove separate hash-set Set<T> implementation from expressions

### PR DESCRIPTION
**Pull request by JonHanna**
Expressions contains a hash-set based on a `Dictionary‎<TKey, ‎TValue>` with unused values. Since `System.Collections.‎Generic.HashSet<T>` is based on the same approach as `Dictionary<TKey, TValue>`, use it instead.

Use the boolean value returned from `HashSet<T>.Add()` to avoid duplicate lookups.

Don't create arrays solely to add elements to new sets.

(The one downside is that `Set<T>` could take a capacity in a constructor while `HashSet‎<T>` cannot).

## Tighten type of `MergeScopes`

Change from set of objects to set of `BlockExpression`; the only type ever added.

Allows for better type-safety, and we can take advantage of that in calling `scope.Variables` directly, rather than going through `GetVariables()`.

## Don't create list for `CompilerScope.GetVariables()`

It's only ever enumerated immediately, with none of the other members of `IList` used, so yield each variable from each merged scope, rather than building up a list.

### CODE REVIEW

**Comment by JonHanna** [comment]
Test Innerloop OSX Debug Build and Test
Test Innerloop OSX Release Build and Test
(java.nio.file.FileAlreadyExistsException: /jenkins/jobs/dotnet_corefx/jobs/osx_debug_prtest/lastSuccessful)

**Comment by JonHanna** [comment]
Test Innerloop CentOS7.1 Debug Build and Test
Test Innerloop CentOS7.1 Release Build and Test
Test Innerloop Ubuntu Debug Build and Test
Test Innerloop Ubuntu Release Build and Test

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/CompilerScope.cs [hunk 1]
```diff
@@ -462,19 +462,23 @@ private void AllocateLocals(LambdaCompiler lc)
             }
         }
 
-        private IList<ParameterExpression> GetVariables()
```
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/CompilerScope.cs]: looks like we never relied on this to be IList

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/CompilerScope.cs [hunk 2]
```diff
@@ -462,19 +462,23 @@ private void AllocateLocals(LambdaCompiler lc)
...
             {
-                return vars;
+                yield return param;
             }
-            var list = new List<ParameterExpression>(vars);
+
             foreach (var scope in MergedScopes)
             {
-                list.AddRange(GetVariables(scope));
+                foreach (ParameterExpression param in scope.Variables)
```
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/CompilerScope.cs]: individual scopes tend to be small. often just one variable. the change makes sense

**Comment by VSadov** [comment]
At one point System.Core was supposed to compile with pre-4.0 dependencies. So I guess that is why it had its own Set implementation

**Comment by VSadov** [comment]
LGTM

**Pull request merged**